docs/src/charmil_config.md

```diff
@@ -0,0 +1,170 @@
+# Charmil Config
+
+The Charmil Config package offers a convenient mechanism for both host and plugin developers to manage configurations in their command-line interface (CLI) applications made using [Cobra](https://github.com/spf13/cobra).
+
+## Features:
+
+- Uses [Viper](https://github.com/spf13/viper) under the hood.
+- Helps in maintaining all available configurations in a single central-local config file.
+- Provides the plugin developers with a set of methods to maintain a configuration map and export it to the host CLI.
+- Provides the host CLI developers with a set of methods to import plugin configuration maps and read/write configurations from/to a local config file.
+
+## Steps to Use:
+
+- ### For host CLI developers:
+
+  1.  Open the file where the root command of your CLI is defined using Cobra.
+
+  2.  Import the Charmil Config package by adding the following line at the top of that file:
+      ```go
+      import "github.com/aerogear/charmil/core/config"
+      ```
+  3.  Store a new instance of the Charmil Config handler by calling the `New` function.
+
+      _Example:_
+
+      ```go
+      h := config.New()
+      ```
+
+  4.  Create an instance of the `File` struct, passing all the required settings for the local config file as fields. Once this is done, pass that `File` instance as an argument into the `InitFile` method, in order to link the local config file to the pointer receiver ie. handler.
+
+      _Example:_
+
+      ```go
+      f = config.File{
+      	Name: "config",
+      	Type: "yaml",
+      	Path: "./examples/host",
+      }
+
+      h.InitFile(f)
+      ```
+
+  5.  Load config values from the local config file using the `Load` method.
+
+      _Example:_
+
+      ```go
+      err := h.Load()
+      if err != nil {
+      	log.Fatal(err)
+      }
+      ```
+
+  6.  Use the `SetValue` and `GetValue` methods respectively to set and get values to/from the current config.
+
+      _Example:_
+
+      ```go
+      h.SetValue("key0", "val0")
+
+      val, err := h.GetValue("key0") // returns "val0"
+      if err != nil {
+      	log.Fatal(err)
+      }
+      ```
+
+  7.  Map a plugin name to its config map using the `SetPluginCfg` method by passing the name of the plugin (as you want it in the config file) as the first argument and the config map imported from the plugin, as the second argument.
+
+      _Example:_
+
+      ```go
+      h.SetPluginCfg("pluginName", pluginCfg)
+      ```
+
+  8.  Call the `MergePluginCfg` method to store the config of every plugin (ie. mapped using the `SetPluginCfg` method) into the current config.
+
+      _Example:_
+
+      ```go
+      h.MergePluginCfg()
+      ```
+
+  9.  Write current config into the local config file using the `Save` method.
```

Comment:
So this is approach where we migrate one time only config..

I do see it too much of hassle.

Lets put all things in the table:

1) some clis hqve their own config and will work with it using structure - like rhoas cli does

2) then cli can become plugin and have the same form of config but saved to the single file

For simplicity we assume that plugin cli was never installed on os and only host is.

What I do not get is - why host needs to interact with plugin and use non typed api (strings).

My understanding was that all we need to do is to provide dynamic config location for fetching config.

So config of plugin can be:

- entire file on it's own
- some subpath of json when running in host

Missing anything?

> Typed by one hand while feeding newborn. Sorry for typos

N

Reply:
Since plugin is inside host we can also import config structures into host.

Reply:
Can you reply to this?

Reply:
Okay so I revisited the Rhoas config implementation and I think now I understand what the issues with my current approach are. Will change all the things that you've mentioned.

I've tried to summarize all the current issues with reasons and solutions for the same, through this comment: https://github.com/aerogear/charmil/pull/108#issuecomment-873180768

Please have a look and let me know if there is anything to be added/removed :)